test/jdk/java/net/httpclient/DebugLoggerTest.java

```diff
@@ -0,0 +1,321 @@
+        Set<Destination> dest = new HashSet<>(3);
+        for (var v : values) {
+            v = v.trim().toLowerCase(Locale.ROOT);
+            if ("log".equals(v)) dest.add(Destination.LOG);
+            if ("err".equals(v)) dest.add(Destination.ERR);
+            if ("out".equals(v)) dest.add(Destination.OUT);
+            if ("true".equals(v)) dest.addAll(EnumSet.of(Destination.ERR, Destination.LOG));
+        }
+        return Set.copyOf(dest);
+    }
+
+    public static void main(String[] args) {
+        err =  new RecordingPrintStream(stdErr, true, UTF_8);
+        out =  new RecordingPrintStream(stdOut, true, UTF_8);
+        System.setErr(err);
+        System.setOut(out);
+        TestHandler logHandler = new TestHandler();
+        Logger julLogger = Logger.getLogger(LOGGER_NAME);
+        julLogger.setLevel(Level.ALL);
+        julLogger.setUseParentHandlers(false);
+        julLogger.addHandler(logHandler);
+        System.Logger sysLogger = System.getLogger(LOGGER_NAME);
+        var debug = Utils.getDebugLogger(() -> "DebugLoggerTest", Utils.DEBUG);
+        String prop = System.getProperty(LOGGER_NAME);
+        stdOut.printf("DebugLoggerTest: [\"%s\", %s] start%n", prop, Arrays.asList(args));
+        var dest = getDestinations(prop);
+        var dest2 = Stream.of(args)
+                .map(Destination::valueOf)
+                .collect(Collectors.toSet());
+        assertEquals(debug.on(), !dest.isEmpty());
+        assertEquals(dest, dest2);
+
+        Predicate<LogRecord> matcher1 = (r) -> r.getMessage() != null && r.getMessage().contains(MESSAGE);
+        doTest(() -> debug.log(MESSAGE), logHandler, dest, MESSAGE, matcher1);
+        Exception thrown = new Exception(MESSAGE3);
+        Predicate<LogRecord> matcher2 = (r) -> r.getMessage() != null
+                && r.getMessage().contains(MESSAGE2)
+                && thrown.equals(r.getThrown());
+        doTest(() -> debug.log(MESSAGE2, thrown), logHandler, dest, MESSAGE2, matcher2);
+        stdOut.printf("Test [\"%s\", %s] passed%n", prop, Arrays.asList(args));
+    }
+
+    private static void doTest(Runnable test,
+                               TestHandler logHandler,
+                               Set<Destination> dest,
+                               String msg,
+                               Predicate<LogRecord> logMatcher) {
+        logHandler.logs.clear();
+        err.drainRecordedData();
+        out.drainRecordedData();
+
+        err.startRecording();
+        out.startRecording();
+        test.run();
+        err.stopRecording();
+        out.stopRecording();
+        String outStr = out.drainRecordedData();
+        String errStr = err.drainRecordedData();
+        List<LogRecord> logs = logHandler.logs.stream().toList();
+
+        if (dest.contains(Destination.ERR)) {
+            if (!errStr.contains(msg)) {
+                throw new AssertionError("stderr does not contain the expected message");
```

Comment:
Do you think we should print the errStr/outStr/logs in these assertion failure messages, to help debug any failures?

Reply:
I believe the stack trace should be enough: the line number should reveal which message was not found.